File: TaskManagementApp/Models/TaskModel.cs
using System;
using Google.Cloud.Firestore;

namespace TaskManagementApp.Models
{
    [FirestoreData]
    public class TaskModel
    {
        [FirestoreDocumentId]
        public string Id { get; set; } = string.Empty;

        [FirestoreProperty]
        public string UserId { get; set; } = string.Empty;

        [FirestoreProperty]
        public string Title { get; set; } = string.Empty;

        [FirestoreProperty]
        public string Description { get; set; } = string.Empty;

        [FirestoreProperty]
        public DateTime DueDate { get; set; }

        [FirestoreProperty]
        public bool IsCompleted { get; set; }

        // Store as string in Firestore, but use enum for internal logic
        [FirestoreProperty]
        public string Status { get; set; } = TaskStatus.Todo.ToString(); // Default value "Todo"

        // Convert string to enum for internal logic
        public TaskStatus StatusEnum
        {
            get => Enum.TryParse(Status, out TaskStatus status) ? status : TaskStatus.Todo; // Default to Todo if invalid
            set => Status = value.ToString(); // Set as string for Firestore
        }

        // Store as string in Firestore, but use enum for internal logic
        [FirestoreProperty]
        public string Priority { get; set; } = TaskPriority.Medium.ToString(); // Default value "Medium"

        // Convert string to enum for internal logic
        public TaskPriority PriorityEnum
        {
            get => Enum.TryParse(Priority, out TaskPriority priority) ? priority : TaskPriority.Medium; // Default to Medium if invalid
            set => Priority = value.ToString(); // Set as string for Firestore
        }
    }

    public enum TaskStatus
    {
        Expired,
        Todo,
        Doing,
        Done
    }

    public enum TaskPriority
    {
        Low,
        Medium,
        High
    }
}

File: TaskManagementApp_Nodejs/models/TaskModel.js
const admin = require('firebase-admin');
const { Firestore } = require('@google-cloud/firestore');

const firestore = new Firestore();

class TaskModel {
  constructor({
    id = '',
    userId = '',
    title = '',
    description = '',
    dueDate = { _seconds: 0, _nanoseconds: 0 },
    isCompleted = false,
    status = TaskStatus.Todo,
    statusEnum = 1,
    priority = TaskPriority.Medium,
    focusTime = 0, // Add focusTime field with default 0 seconds
    focusSessions = [], // Array of {date: timestamp, duration: seconds}
  } = {}) {
    this.id = id;
    this.userId = userId;
    this.title = title;
    this.description = description;
    this.dueDate = dueDate;
    this.isCompleted = isCompleted;
    this.status = status;
    this.statusEnum = statusEnum;
    this.priority = priority;
    this.focusTime = focusTime;
    this.focusSessions = focusSessions;
  }

  static fromFirestore(doc) {
    const data = doc.data();
    return new TaskModel({
      id: doc.id,
      userId: data.UserId,
      title: data.Title,
      description: data.Description,
      dueDate: data.DueDate,
      isCompleted: data.IsCompleted,
      status: data.Status,
      statusEnum: data.StatusEnum,
      priority: data.Priority,
      focusTime: data.FocusTime || 0, // Get focusTime from Firestore or default to 0
      focusSessions: data.FocusSessions || [],
    });
  }

  toFirestore() {
    return {
      UserId: this.userId,
      Title: this.title,
      Description: this.description,
      DueDate: this.dueDate,
      IsCompleted: this.isCompleted,
      Status: this.status,
      StatusEnum: this.statusEnum,
      Priority: this.priority,
      FocusTime: this.focusTime, // Save focusTime to Firestore
      FocusSessions: this.focusSessions,
    };
  }
}

const TaskStatus = {
  Expired: 'Expired',
  Todo: 'Todo',
  Doing: 'Doing',
  Done: 'Done',
};

const TaskPriority = {
  Low: 'Low',
  Medium: 'Medium',
  High: 'High',
};

module.exports = { TaskModel, TaskStatus, TaskPriority };